librz/include/rz_search.h

```diff
@@ -12,117 +12,46 @@ extern "C" {
-RZ_API int rz_search_magic_update(RzSearch *_s, ut64 from, const ut8 *buf, int len);
-RZ_API int rz_search_deltakey_update(RzSearch *s, ut64 from, const ut8 *buf, int len);
-RZ_API int rz_search_strings_update(RzSearch *s, ut64 from, const ut8 *buf, int len);
-RZ_API int rz_search_regexp_update(RzSearch *s, ut64 from, const ut8 *buf, int len);
-// Returns 2 if search.maxhits is reached, 0 on error, otherwise 1
-RZ_API int rz_search_hit_new(RzSearch *s, RzSearchKeyword *kw, ut64 addr);
-RZ_API void rz_search_set_distance(RzSearch *s, int dist);
-RZ_API int rz_search_set_string_limits(RzSearch *s, ut32 min, ut32 max); // dup again?
-// RZ_API int rz_search_set_callback(RzSearch *s, int (*callback)(struct rz_search_kw_t *, void *, ut64), void *user);
-RZ_API void rz_search_set_callback(RzSearch *s, RzSearchCallback(callback), void *user);
-RZ_API int rz_search_begin(RzSearch *s);
-
-/* pattern search */
-RZ_API void rz_search_pattern_size(RzSearch *s, int size);
-RZ_API int rz_search_pattern(RzSearch *s, ut64 from, ut64 to);
+typedef bool (*RzSearchCancelCallback)(void *user, size_t n_hits);
+
+RZ_API bool rz_search_opt_set_backwards(RZ_NONNULL RzSearchOpt *opt, bool backwards);
+RZ_API bool rz_search_opt_set_allow_overlaps(RZ_NONNULL RzSearchOpt *opt, bool allow_overlaps);
+RZ_API bool rz_search_opt_set_inverse_match(RZ_NONNULL RzSearchOpt *opt, bool inverse_match);
+RZ_API bool rz_search_opt_set_buffer_size(RZ_NONNULL RzSearchOpt *opt, size_t buffer_size);
+RZ_API bool rz_search_opt_set_max_threads(RZ_NONNULL RzSearchOpt *opt, size_t max_threads);
+RZ_API bool rz_search_opt_set_cancel_cb(RZ_NONNULL RzSearchOpt *opt, RzSearchCancelCallback callback, void *user);
+
+RZ_API RZ_OWN RzSearchCollection *rz_search_collection_aes_keys();
+
+RZ_API RZ_OWN RzSearchCollection *rz_search_collection_private_keys();
+
+RZ_API RZ_OWN RzSearchCollection *rz_search_collection_regex();
+RZ_API bool rz_search_collection_regex_add(RZ_NONNULL RzSearchCollection *col, RZ_NONNULL const char *regex, bool caseless);
+
+RZ_API RZ_OWN RzSearchCollection *rz_search_collection_bytes();
+RZ_API bool rz_search_collection_bytes_add(RZ_NONNULL RzSearchCollection *col, RZ_NONNULL const char *metadata, RZ_NONNULL const ut8 *bytes, RZ_NULLABLE const ut8 *mask, size_t length);
+RZ_API bool rz_search_collection_bytes_add_pattern(RZ_NONNULL RzSearchCollection *col, RZ_NONNULL const char *hex_pattern);
+
+RZ_API RZ_OWN RzSearchCollection *rz_search_collection_magic(RZ_NONNULL const char *magic_dir);
+
+RZ_API bool rz_search_collection_match_any(RZ_NULLABLE RzSearchCollection *sc, RZ_NONNULL const ut8 *buffer, size_t length);
+RZ_API void rz_search_collection_free(RZ_NULLABLE RzSearchCollection *sc);
+RZ_API void rz_search_hit_free(RZ_NULLABLE RzSearchHit *hit);
+
+RZ_API RZ_OWN RzList /*<RzSearchHit *>*/ *rz_search_run(RZ_NONNULL RzSearchOpt *opt, RZ_NONNULL RzSearchCollection *col, RZ_NONNULL RzIO *io, RZ_NONNULL RzList /*<RzIOMap *>*/ *search_in);
```

Comment:
Maybe changing the `search_in` parameter to `void *` and infer the type based on some flag in `opt`. This way we can use this function later to search other things then IO.

Reply:
Not so sure about this, but maybe.